kirkwood: move usb/sd/sata support from built-in modules to kmod packages

CHKDSK88 opened the pull request:
This PR moves usb/sd/sata support from built-in modules to kmod packages. Those modules can be used as kmods, so we don't need it direct in uImage.

Tested-on: pogo v4

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
kirkwood: move sata features to modules

All devices are using nand images. Built-in SATA modules are not needed
anymore.

Tested-by: Cezary Jackiewicz <cezary@eko.one.pl> [nsa310]
Tested-by: Alberto Bursi <bobafetthotmail@gmail.com> [nsa325, ea4500]
Reviewed-by: Sungbo Eo <mans0n@gorani.run>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
kirkwood: move mmc/sd features to modules

All devices are using nand images. Built-in MMC/SD modules are not needed
anymore.

Run tested: pogo v4

Reviewed-by: Sungbo Eo <mans0n@gorani.run>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
kirkwood: move usb support to modules

USB support built in kernel isn't required for booting.
This patch move USB support to modules.

CONFIG_USB_STORAGE is removed because it can be find in kmod-usb-storage
and all usb support for this targed is moved to kmods.

CONFIG_SCSI is removed because it can be find in kmod-scsi-core, which
is selected if it's needed.

Tested-by: Cezary Jackiewicz <cezary@eko.one.pl> [nsa310]
Tested-by: Alberto Bursi <bobafetthotmail@gmail.com> [nsa325, ea4500]
Reviewed-by: Sungbo Eo <mans0n@gorani.run>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
Revert "kirkwood: disable image for Linksys EA3500 (Audi)"

This reverts commit d9ff49967145600ff4bbc0f4eecef11fd96f6b16.

At this moment kernel can be fitted in EA3500 image.

Tested by onefreeman (OpenWrt forum user)
Tested-by: Jacks <jack338c@gmail.com> (Github user @jack338c)
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

mans0n commented:
IMHO it would be better to put kmod-mvsdio into kirkwood/module.mk. See https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/modules.mk#L29.

For USB_LED_TRIG, I think we can just remove it. We have kmod-usb-ledtrig-usbport as an alternative, and no device is using usb-host trigger anyway.

dengqf6 commented on target/linux/kirkwood/config-5.4:
```diff
@@ -120,7 +118,6 @@ CONFIG_DTC=y
 # CONFIG_EARLY_PRINTK is not set
 CONFIG_EDAC_ATOMIC_SCRUB=y
 CONFIG_EDAC_SUPPORT=y
-CONFIG_EXT4_FS=y
```
EXT4 depends on JBD2, which can be removed as well

CHKDSK88 replied:
Done. Thank You.

CHKDSK88 commented:
@mans0n 

>IMHO it would be better to put kmod-mvsdio into kirkwood/module.mk. See https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/modules.mk#L29.

kmod-mvsdio could be used as well in mvebu, in armada 375.

>For USB_LED_TRIG, I think we can just remove it. We have kmod-usb-ledtrig-usbport as an alternative, and no device is using usb-host trigger anyway.

It is used by Zyxel nsa, see 02_leds. If someone could test, I can rework it to kmod-usb-ledtrig-usbport.

mans0n commented:
> kmod-mvsdio could be used as well in mvebu, in armada 375.

Ah, you're right. Let's keep it there then.

> It is used by Zyxel nsa, see 02_leds.

Oops, sorry, I missed that. But it still does not make sense, as usb-ledtrig is not a separate module, it is a part of usb-common.

adrianschmutzler commented:
If you sorted it out, feel free to provided Reviewed-by statements :-)

obsy commented:
Tested on zyxel nsa-310

Tested-by: Cezary Jackiewicz <cezary@eko.one.pl>

mans0n commented:
@obsy Hi, could you also run `lsusb -t` on your device? (with usbutils installed and USB inserted)

obsy commented:
```
root@OpenWrt:~# lsusb -t
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=orion-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
        |__ Port 2: Dev 3, If 0, Class=Mass Storage, Driver=usb-storage, 480M

```

bobafetthotmail commented:
Tested on NSA310b and Linksys EA4500.

All good, split Sata leds for each port work fine.
Sata and USB drives can be accessed in both Sata/eSata port and USB.

Tested-by: Alberto Bursi <bobafetthotmail@gmail.com>

Some feedback:
-you are adding kmod-ata-marvell-sata kmod-ata-ledtrig packages to default target packages, the two linksys routers don't need them.

- the ata-sata-pmp package should be added to all devices that have Sata ports (it's the port multiplier support)

-since this target had ext4 support in kernel since forever, I think you should add ext4 kmod to default packages, so all devices remain able to use ext4 by default and nothing changes for users.

> If someone could test, I can rework it to kmod-usb-ledtrig-usbport.

I can test this

CHKDSK88 committed:
kirkwood: move led triggers to dts

This patch moves led trigger settings from 01_leds file to dts.

Now is used "linux,default-trigger" property.

Tested-by: Alberto Bursi <bobafetthotmail@gmail.com> [nsa310,nsa325]
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

mans0n commented:
Tested on an unofficial device. sata, usb, and led work well.

Dummy packages do not have module files, so we don't need to use AUTOLOAD for them.
And try removing unnecessary configs with `kernel_oldconfig`.

> the ata-sata-pmp package should be added to all devices that have Sata ports (it's the port multiplier support)

If sata-pmp is that necessary, then how about adding the config to ata-core?

@obsy @bobafetthotmail Could you try this? (with kmod-usb-ledtrig-usbport installed)

`ucidef_set_led_usbhost "usb" "USB" "nsa310:green:usb"`
->
`ucidef_set_led_usbport "usb" "USB" "nsa310:green:usb" "1-1-port2"`

It seems nsa325 has two usb ports. I'm not sure about that yet...

obsy commented:
Already tested. For nsa-310: front: 1-1-port2 i back: 1-1-port3

mans0n commented:
Thanks. I've missed the front port...

mans0n commented:
If I enable kmod-usb-ledtrig, USB_COMMON becomes y, not m. USB_COMMON does not depend on USB, so I think we can just put it in `target/linux/kirkwood/config-5.4`.

bobafetthotmail commented:
> If sata-pmp is that necessary, then how about adding the config to ata-core?
> 

You can see more discussion about what PMP is and what it does in the PR where I added support to kirkwood https://github.com/openwrt/openwrt/pull/2007

If you move it to a dummy package like that, I have to recompile the kernel of the device if I want it, and that's bad.

If you keep it as a kmod module that I can install then it's fine even if it's not installed by default.

As for adding to ata-core, afaik port multiplier support matters mostly for devices like kirkwood and mvebu that can't just add a PCIe card with more Sata ports, for x86 (PC/server) the port multiplier is not really a thing anymore.

There is also a PR where someone added PMP support to mvebu https://github.com/openwrt/openwrt/pull/1021 and none had any complaint whatsoever, while when I did the same for Kirkwood everyone had to complain, because apparently I'm special like that.

> It seems nsa325 has three usb ports. I'm not sure about that yet...

The nsa325 ports on the back are USB 2.0 from a hub, afaik the Kirkwood SoC has a single USB 2.0 controller.

The port on the front is a USB 3.0 from a controller attached to the SoC PCIe lanes.

I have a NSA325, I suppose you want a `lsusb -t` on it with this patch?

mans0n commented:
Thanks for the references.

> As for adding to ata-core, afaik port multiplier support matters mostly for devices like kirkwood and mvebu that can't just add a PCIe card with more Sata ports, for x86 (PC/server) the port multiplier is not really a thing anymore.

Hmm, it seems Alpine Linux also enables SATA_PMP. If we pull libata out of built-in kernel, then sata-pmp will not take more kernel partition space, and I don't really think a device having sata port would have insufficient roofs space. I think it's okay to add SATA_PMP to kmod-ata. Maybe we should open another PR for it...

> I have a NSA325, I suppose you want to see a lsusb -t on it with this patch?

Oh, that's great. Please let us know. :)

bobafetthotmail commented:
> Hmm, it seems Alpine Linux also enables SATA_PMP. 

Afaik all decent Linux distros enable SATA_PMP.

I know for a fact that Debian/Ubuntu and OpenSUSE do because I did try port multipliers on AMD systems less than a year ago.

Not all hardware on x86 supports PMP, for example Intel Sata controllers don't support it. But it is not an issue.

> I think it's okay to add SATA_PMP to kmod-ata. Maybe we should open another PR for it...

I also think it is OK, if the hardware does not support it there are no ill effects, as said above.
I'm just worried about all the OpenWrt members coming and complaining for no reason like it happened when I added PMP support to kirkwood.

> Oh, that's great. Please let us know. :)

This is with nothing attached (first two are the same port, in USB 3.0 and 2.0 mode respectively, the last bus is the native USB 2.0 controller, and the last line is the onboard hub to provide more than 1 port on the back panel, as I said above)
```
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 5000M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=orion-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M

```
Now with a USB 2.0 flash drive attached to the USB 3.0 port
```
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 5000M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 480M
    |__ Port 1: Dev 2, If 0, Class=Mass Storage, Driver=, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=orion-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
```

Now with a USB 3.0 flash drive attached to USB 3.0 port
```

/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 5000M
    |__ Port 1: Dev 2, If 0, Class=Mass Storage, Driver=, 5000M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=orion-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
```


Now with the USB 2.0 flashdrive attached to the USB 2.0 port called "1" on the back panel
```
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 5000M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=orion-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
        |__ Port 2: Dev 3, If 0, Class=Mass Storage, Driver=, 480M
```

now on the other port, called "2"
```
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 5000M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 480M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=orion-ehci/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
        |__ Port 3: Dev 4, If 0, Class=Mass Storage, Driver=, 480M

```

CHKDSK88 committed:
kirkwood: refresh kernel config

It was made by running "make kernel_oldconfig".

Suggested-by: Sungbo Eo <mans0n@gorani.run>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
kernel: add sata multiplier to ata-core package

At this moment this option is enabled only in kirkwood and mvebu.
Adding it to generic kmod-ata-core allow other platforms users to
use SATA multipliers.

Suggested-by: Alberto Bursi <bobafetthotmail@gmail.com>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 commented:
Thank You for tests and feedback.

I added some changes and some was reworked. Config is refreshed, led triggers moved to dts. I added commit with push PMP to ata-core package. This is separate commit, because I'm not sure if it will be merged.

@obsy convinced me, that usb should stay as host trigger. This trigger give info about traffic and that is important for USB leds in this case. We can't make this led trigger properly.

The last issue:
Should I add ext packages to NAS devices? 
@adrianschmutzler @mans0n , what is Your opinion?

adrianschmutzler commented:
@hauke Can you have a look at the kernel built-in to package changes here, whether they are good to go?

bobafetthotmail commented:
> Should I add ext packages to NAS devices?
I would add it to the whole target, as before it was enabled for every Kirkwood. It's more to keep the same situation as before than because they need it.
It's not a 100% requirement, I'm fine either way.

I'll run test NSA310 and NSA325 for the led change.

bobafetthotmail commented:
Ok if you want you can add my tested by to **kirkwood: move led triggers to dts**  for both nsa310 and nsa325

Tested-by: Alberto Bursi <bobafetthotmail@gmail.com>

mans0n commented:
> Should I add ext packages to NAS devices?

I'm okay with both. If we are going to add it, then I would personally prefer to make the whole ext4 stuff a separate commit.

I suggested refreshing config because some configs such as BLK_DEV_SD and FS_IOMAP become unnecessary after removing USB and EXT4_FS. Try refreshing config for every move commit.

And could you please consider keeping USB_LED_TRIG in config-5.4? All target modules will be enabled by buildbot, and there's no way to keep USB_COMMON as a module. kmod-ata-ledtrig will increase the size of kmod-ata, but kmod-usb-ledtrig will increase the kernel itself, so it will not help for our objective, decreasing the kernel size.

CHKDSK88 commented:
@mans0n 

>I'm okay with both. If we are going to add it, then I would personally prefer to make the whole ext4 stuff a separate commit.

Good idea. 
And could you please consider keeping USB_LED_TRIG in config-5.4? All target modules will be enabled by buildbot, and there's no way to keep USB_COMMON as a module. kmod-ata-ledtrig will increase the size of kmod-ata, but kmod-usb-ledtrig will increase the kernel itself, so it will not help for our objective, decreasing the kernel size.
I will prepare patch for NAS devices. Linksyses don't need ext4.

>I suggested refreshing config because some configs such as BLK_DEV_SD and FS_IOMAP become unnecessary after removing USB and EXT4_FS. Try refreshing config for every move commit.

Thank You for advice. I will put config refresh at the end.

>And could you please consider keeping USB_LED_TRIG in config-5.4? All target modules will be enabled by buildbot, and there's no way to keep USB_COMMON as a module. kmod-ata-ledtrig will increase the size of kmod-ata, but kmod-usb-ledtrig will increase the kernel itself, so it will not help for our objective, decreasing the kernel size.

Please look:
https://github.com/torvalds/linux/blob/219d54332a09e8d8741c1e1982f5eae56099de85/drivers/usb/common/Makefile#L9

Leds are part of usb-common, which is part of usb-core: https://github.com/openwrt/openwrt/blob/633c5137df86e782759e0d88c0facfb2d3f0c424/package/kernel/linux/modules/usb.mk#L21

That is a reason why i made dummy package.

CHKDSK88 committed:
kirkwood: move ext4 support to modules

Ext4 support is built in kirkwood kernel.

This patch moves ext4 support to modules and enable it for devices
with SATA interface.

Reviewed-by: Sungbo Eo <mans0n@gorani.run>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

mans0n commented:
> Leds are part of usb-common, which is part of usb-core: 

Sorry, I didn't noticed that. But what I wanted to say is that USB_LED_TRIG is bool (not tristate) and USB_COMMON will become y if USB_LED_TRIG is y. That's why I said "there's no way to keep USB_COMMON as a module", the existence of kmod-usb-ledtrig itself will make usb-common built-in. And USB_COMMON does not depend on USB (only USB _selects_ USB_COMMON), so we can enable USB_COMMON without enabling USB.
But I admit enabling USB_COMMON without USB is a little bit strange, so I won't insist it anymore if you still prefer dummy-package way.

CHKDSK88 commented:
>the existence of kmod-usb-ledtrig itself will make usb-common built-in (in case of buildbot). 

You're right. I will rework it.

mans0n commented on target/linux/kirkwood/image/Makefile:
```diff
@@ -50,7 +50,8 @@ define Device/cloudengines_pogoplugv4
   DEVICE_MODEL := Pogoplug V4
   DEVICE_DTS := kirkwood-pogoplug-series-4
-  DEVICE_PACKAGES := kmod-usb3
+  DEVICE_PACKAGES := kmod-ata-marvell-sata kmod-ata-ledtrig ata-sata-pmp \
```
`kmod-ata-sata-pmp`?

mans0n commented:
> This patch moves led trigger settings from 02_leds file to dts.

It should be `01_leds`.

CHKDSK88 commented:
@mans0n @bobafetthotmail @adrianschmutzler 

I hope, that now everything is ok.

mans0n commented:
You forgot to remove kmod-ata-sata-pmp from nsa325.
If `CONFIG_USB=y` is added, then `# CONFIG_USB_EHCI_HCD is not set` might be needed as well. (Not sure, but kernel_oldconfig prompted it)
With that, `Reviewed-by: Sungbo Eo <mans0n@gorani.run>` for the four move patches.

Thanks for your hard work!

CHKDSK88 commented:
@mans0n 

I put ```# CONFIG_USB_EHCI_HCD is not set``` to generic config. R removed kmod-ata-sata-pmp.

>Thanks for your hard work!

Thank You for Your time and help. :)

CHKDSK88 commented:
@jack338c

If You test EA3500 with this PR, I would like put "Tested-By" to commit message. If You ok with that,  need Your real name and e-mail.

jack338c commented:
@CHKDSK88 
ok,my name Jacks
email: jack338c@gmail.com
Thanks

CHKDSK88 commented:
@adrianschmutzler 

Is the blocker patch "kernel: add sata multiplier to ata-core package" only?

adrianschmutzler commented:
@CHKDSK88 No, the blocker is that my expertise is limited on high-level stuff only, and if I wanted to review the patches beyond just believing you it would require me a substantial amount of time.

However, I believe that other committers could do this with much less effort, so for now I wait for somebody else to review.

adrianschmutzler commented on target/linux/kirkwood/modules.mk:
```diff
@@ -0,0 +1,32 @@
+
+
+define KernelPackage/ata-sata-pmp
+  SUBMENU:=$(BLOCK_MENU)
```
This is unneeded, since the AddDepends will set it anyway?

adrianschmutzler commented on package/kernel/linux/modules/other.mk:
```diff
@@ -373,6 +373,23 @@ endef
 $(eval $(call KernelPackage,mmc))
 
 
+define KernelPackage/mvsdio
```
I wonder whether calling this mmc-mvsdio would be better?

adrianschmutzler replied:
never mind, just read the docs ...

adrianschmutzler commented on package/kernel/linux/modules/other.mk:
```diff
@@ -373,6 +373,23 @@ endef
+define KernelPackage/mvsdio
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=Marvell MMC/SD/SDIO host driver
+  DEPENDS:=+kmod-mmc @TARGET_kirkwood
```
Why is this added here and not in target modules.mk? Do you expect it to be reused by other target?

adrianschmutzler replied:
Found the comment about this as well ...

adrianschmutzler commented on target/linux/kirkwood/config-5.4:
```diff
@@ -339,12 +338,7 @@ CONFIG_UNWINDER_ARM=y
-CONFIG_USB_EHCI_HCD_ORION=y
-# CONFIG_USB_EHCI_HCD_PLATFORM is not set
 CONFIG_USB_LED_TRIG=y
-CONFIG_USB_STORAGE=y
```
Can you comment on why CONFIG_SCSI and CONFIG_USB_STORAGE are removed here?

adrianschmutzler commented:
I consider merging the first 6 patches, I suppose not having the last two won't be a problem for the moment?

Despite, off-topic, I consider to rename audi and viper to their product model names as well. Opinions?

adrianschmutzler commented:
And some of the tested-by addresses lack the brackets for the mail addresses.

adrianschmutzler commented on target/linux/kirkwood/modules.mk:
```diff
@@ -0,0 +1,32 @@
+define KernelPackage/ata-ledtrig
+  TITLE:=LED trigger for ATA port
+  DEPENDS:=@TARGET_kirkwood
+  KCONFIG:=CONFIG_ATA_LEDS=y
```
If this is put here with =y, how will this be different from having it set in config-5.4 directly?

CHKDSK88 replied:
Because it is depended on ata-core and every kernel refresh will remove it from config.

CHKDSK88 commented:
@adrianschmutzler 

PR rebased to current master. All changes done.

>I consider merging the first 6 patches, I suppose not having the last two won't be a problem for the moment?

You mean last in github order or git order?

I know that "kernel: add sata multiplier to ata-core package" is problematic, thats the reason why I put it to separate commit. But which second?
I will send controversial commits  to ML after merging rest.

>Despite, off-topic, I consider to rename audi and viper to their product model names as well. Opinions?

Kirkwood could be harmonized with mvebu. So I can prepare PR with changed model names and port names in dsa part of dts after this PR. Is that ok for You?

adrianschmutzler commented:
These I plan to merge:
kirkwood: move led triggers to dts …
kirkwood: move sata features to modules …
kirkwood: move mmc/sd features to modules …
kirkwood: move usb support to modules …
Revert "kirkwood: disable image for Linksys EA3500 (Audi)" …
kirkwood: refresh kernel config …

These I plan to leave out for now:
kernel: add sata multiplier to ata-core package …
kirkwood: move ext4 support to modules …

> Kirkwood could be harmonized with mvebu. So I can prepare PR with changed model names and port names in dsa part of dts after this PR. Is that ok for You?

I already tried that yesterday here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/kirkwoodrename

However, looking at it, I'm not sure anymore whether this will really be an "improvement", or just making things bad in a different way.

CHKDSK88 commented:
> kirkwood: move ext4 support to modules 

Could You tell why not this patch?

>However, looking at it, I'm not sure anymore whether this will really be an "improvement", or just making things bad in a different way.

I'm not sure if after commit is better.

adrianschmutzler commented:
And what about CONFIG_SCSI and CONFIG_USB_STORAGE, I don't see where those are implemented in the modules, or are they dropped entirely by intention?

adrianschmutzler commented:
> And what about CONFIG_SCSI and CONFIG_USB_STORAGE,

@CHKDSK88 Just saw you put that in the commit message :-)
Will build-test now ...

adrianschmutzler commented:
Also did a make kernel_oldconfig:

```
scripts/kconfig/conf  --oldconfig Kconfig
net/sched/Kconfig:45: warning: menuconfig statement without prompt

WARNING: unmet direct dependencies detected for ARCH_WANT_LIBATA_LEDS
  Depends on [n]: ATA [=n]
  Selected by [y]:
  - MACH_KIRKWOOD [=y] && ARCH_MVEBU [=y] && ARCH_MULTI_V5 [=y]
```

resulting diff:
```
diff --git a/target/linux/kirkwood/config-5.4 b/target/linux/kirkwood/config-5.4
index a5d00a9920..6a81164fb2 100644
--- a/target/linux/kirkwood/config-5.4
+++ b/target/linux/kirkwood/config-5.4
@@ -116,6 +116,8 @@ CONFIG_EXT4_FS=y
 CONFIG_FIXED_PHY=y
 CONFIG_FIX_EARLYCON_MEM=y
 CONFIG_FORCE_PCI=y
+CONFIG_FS_IOMAP=y
+CONFIG_FS_MBCACHE=y
 CONFIG_FW_LOADER_PAGED_BUF=y
 CONFIG_GENERIC_ALLOCATOR=y
 CONFIG_GENERIC_ATOMIC64=y
@@ -138,6 +140,7 @@ CONFIG_GLOB=y
 CONFIG_GPIOLIB=y
 CONFIG_GPIO_MVEBU=y
 CONFIG_GPIO_SYSFS=y
+CONFIG_GRO_CELLS=y
 CONFIG_HANDLE_DOMAIN_IRQ=y
 CONFIG_HARDIRQS_SW_RESEND=y
 CONFIG_HAS_DMA=y
```

adrianschmutzler commented:
Okay, after "kirkwood: move ext4 support to modules" has been added, only `+CONFIG_GRO_CELLS=y` remains.

adrianschmutzler commented on package/kernel/linux/modules/block.mk:
```diff
@@ -26,7 +26,9 @@ define KernelPackage/ata-core
   DEPENDS:=@PCI_SUPPORT||TARGET_sunxi +kmod-scsi-core
-  KCONFIG:=CONFIG_ATA
+  KCONFIG:= \
+	CONFIG_SATA_PMP=y \
```
Since buildbots build with all kmods enabled and this is =y, wouldn't this increase the kernel size for _all_ targets?

adrianschmutzler commented on target/linux/kirkwood/config-5.4:
```diff
@@ -53,9 +53,7 @@ CONFIG_ARM_L1_CACHE_SHIFT=5
 CONFIG_ARM_UNWIND=y
-CONFIG_ATA=y
 CONFIG_ATAGS=y
-CONFIG_ATA_LEDS=y
```
This patch is the only thing that still bugs me.

It doesn't sound really right to me to have CONFIG_ATA_LEDS=y built into the kernel but CONFIG_ATA to be built as module.

Despite, we still have CONFIG_ARCH_WANT_LIBATA_LEDS=y here, which creates the "unmet direct dependencies" warning.

I'd rather say let's keep CONFIG_ATA, CONFIG_ATA_LEDS and CONFIG_SATA_PMP in the config, and only use the module for CONFIG_SATA_MV, where it's straightforward.

mans0n replied:
I think that's also reasonable if CONFIG_ATA does not take so much space.

adrianschmutzler commented:
I've pushed a tree without the mentioned ATA changes here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/kirkwood3

For that, the dependency warning is removed, but additional symbols are added by kernel_oldconfig.

I will start a full build now. Maybe somebody can comment on the new symbols and/or test on device.

bobafetthotmail commented:
> Despite, off-topic, I consider to rename audi and viper to their product model names as well. Opinions?

Agreed. code names are just confusing.

adrianschmutzler commented:
Just tested "my" build from kirkwood3 tree and got 2413 kiB for the audi uImage (with buildbot settings except sdk and imagebuilder).
The partition is 2624k, so that should be enough.

adrianschmutzler commented:
@CHKDSK88 @mans0n @bobafetthotmail @obsy 

I have tidied up my staging branch:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/kirkwood3

The content is exactly the same as for my earlier posting here, but I have now put the changes into the appropriate commits and rearranged the order slightly. I have also removed the Reviewed-by/Tested-by for the sata patch, as I changed too much of the content there to keep them.

I'd be willing to merge the changes in this state, and would be happy about some quick Reviewed-by/Tested-by for the updated version. However, I don't think we need the full range of tests a second time ...

adrianschmutzler commented:
> Despite, off-topic, I consider to rename audi and viper to their product model names as well.

I've sent a patch to the mailing list for that (and dropped the corresponding branch for my staging area)

CHKDSK88 commented:
>Okay, after "kirkwood: move ext4 support to modules" has been added, only +CONFIG_GRO_CELLS=y remains.

Should I take care about that at this moment? Or just test Your staging tree?

>I'd be willing to merge the changes in this state, and would be happy about some quick Reviewed-by/Tested-by for the updated version. However, I don't think we need the full range of tests a second time ...

I will make run test on pogo v4 for sata.

adrianschmutzler commented:
> > Okay, after "kirkwood: move ext4 support to modules" has been added, only +CONFIG_GRO_CELLS=y remains.
> 
> Should I take care about that at this moment? Or just test Your staging tree?

I've done another refresh with make kernel_oldconfig and merged it together with yours in my tree.
I've kept the CONFIG_GRO_CELLS=y added by the process there, as it would be added with the next refresh anyway.
After not removing CONFIG_ATA, it also added back a few of the symbols your kernel_oldconfig removed before that point (e.g. SCSI).

But anyway, I'm quite satisfied with the current result and would be fine with one or two device tests of my tree, unless you think I did something wrong there (i.e. "just test my tree").

> 
> > I'd be willing to merge the changes in this state, and would be happy about some quick Reviewed-by/Tested-by for the updated version. However, I don't think we need the full range of tests a second time ...
> 
> I will make run test on pogo v4 for sata.

Thanks.

After all, this saves > 200k for the kernel (even after my changes), so I think this is a jump forward for future support of this target.

CHKDSK88 commented:
Everything work as expected. Sata log:
```
[    3.529646] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    3.651185] sata_mv f1080000.sata: slots 32 ports 1
[    3.658681] scsi host0: sata_mv
[    3.662250] ata1: SATA max UDMA/133 irq 32
[    4.159572] ata1: SATA link up 3.0 Gbps (SStatus 123 SControl F300)
[    4.166179] ata1.00: ATA-10: 8GB SATA Flash Drive, SFPS921D, max UDMA/133
[    4.172962] ata1.00: 15649200 sectors, multi 16: LBA48 NCQ (depth 31/32)
[    4.180081] ata1.00: configured for UDMA/133
[    4.184790] scsi 0:0:0:0: Direct-Access     ATA      8GB SATA Flash D 921D PQ: 0 ANSI: 5
[    4.194474] sd 0:0:0:0: [sda] 15649200 512-byte logical blocks: (8.01 GB/7.46 GiB)
[    4.202583] sd 0:0:0:0: [sda] Write Protect is off
[    4.207574] sd 0:0:0:0: [sda] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
[    4.220462]  sda: sda1 sda2
[    4.224912] sd 0:0:0:0: [sda] Attached SCSI disk
```
Ata led trigger work. No issues in bootlog. USB work, SD work.

So:
`Tested-by: Pawel Dembicki <paweldembicki@gmail.com>`

We have some place to move in future, when kernel will grown more than 200k. But at this moment, I think, that main goal was reached. Support for Linksys EA3500 is safe now.

adrianschmutzler commented:
Thanks to all contributors for this efforts, I've merged my altered tree now.

If somebody sticks to the idea that removing CONFIG_ATA would be beneficial, feel free to submit this again separately. Personally, I would only consider that as last resort, given that ATA_LEDS needs to be =y.

For "kernel: add sata multiplier to ata-core package" a rewrite would be necessary, as I have left this in target config for now (effectively, the resulting patch would be much easier).
However, if CONFIG_SATA_PMP really would be set up with =y in the common config, the resulting size increase for many targets would be a problem IMO.
If somebody is interested in submitting a rebased version, this should go to the list, and I would recommend to measure kernel size increase on a few targets.

CHKDSK88 commented:
@adrianschmutzler 

Thank you for your time and help.
